File: iosApp/iosApp/Services/Wallet/WalletServiceAssembler.swift
// Created by web3d3v on 03/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

final class WalletServiceAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .singleton) { resolver -> WalletService in
            DefaultWalletService(
                networkService: resolver.resolve(),
                currencyStoreService: resolver.resolve(),
                currenciesCache: KeyValueStore(
                    name: "WalletService.currencies"
                ),
                networksStateCache: KeyValueStore(
                    name: "WalletService.networksState"
                )
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/services/WalletServiceAssembler.kt
package com.sonsofcrypto.web3wallet.android.services

import com.sonsofcrypto.web3lib.keyValueStore.KeyValueStore
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.DefaultWalletService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import smartadapter.internal.extension.name

class WalletServiceAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(WalletService::class.name, AssemblerRegistryScope.SINGLETON) {
            DefaultWalletService(
                it.resolve(NetworksService::class.name),
                it.resolve(CurrencyStoreService::class.name),
                it.resolve("${KeyValueStore::class.name}.WalletService.Currencies"),
                it.resolve("${KeyValueStore::class.name}.WalletService.NetworksState"),
                it.resolve("${KeyValueStore::class.name}.WalletService.TransferLogsCache"),
            )
        }
    }
}